cpp/src/arrow/union_array-test.cc

```diff
@@ -0,0 +1,278 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <string>
+
+#include <gtest/gtest.h>
+
+#include "arrow/compute/test-util.h"
+#include "arrow/testing/gtest_util.h"
+#include "arrow/testing/util.h"
+#include "arrow/type.h"
+
+using arrow::internal::checked_cast;
+using arrow::internal::checked_pointer_cast;
+
+namespace arrow {
+
+class TestUnionArray : public ::testing::Test {
+ public:
+  void SetUp() {
+    pool_ = default_memory_pool();
+    auto type_ids =
+        compute::_MakeArray<Int8Type, int8_t>(int8(), {0, 1, 2, 0, 1, 3, 2, 0, 2, 1}, {});
```

Comment:
You should take a look at `ArrayFromJSON`. It generally makes writing such code easier.